core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -75,152 +90,125 @@ protected boolean localExecute(Request request) {
+                new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        doStart();
+                    }
 
-                @Override
-                public String executor() {
-                    return ThreadPool.Names.SAME;
-                }
+                    @Override
+                    public void onClusterServiceClose() {
+                        listener.onFailure(new NodeClosedException(clusterService.localNode()));
+                    }
 
-                @Override
-                public void handleException(final TransportException exp) {
-                    if (exp.unwrapCause() instanceof ConnectTransportException) {
-                        // we want to retry here a bit to see if a new master is elected
-                        logger.debug("connection exception while trying to forward request to master node [{}], scheduling a retry. Error: [{}]",
-                                nodes.masterNode(), exp.getDetailedMessage());
-                        observer.waitForNextChange(new ClusterStateObserver.Listener() {
-                                                       @Override
-                                                       public void onNewClusterState(ClusterState state) {
-                                                           innerExecute(request, listener, observer, false);
-                                                       }
-
-                                                       @Override
-                                                       public void onClusterServiceClose() {
-                                                           listener.onFailure(new NodeClosedException(clusterService.localNode()));
-                                                       }
-
-                                                       @Override
-                                                       public void onTimeout(TimeValue timeout) {
-                                                           listener.onFailure(new MasterNotDiscoveredException());
-                                                       }
-                                                   }, new ClusterStateObserver.EventPredicate() {
-                                                       @Override
-                                                       public boolean apply(ClusterChangedEvent event) {
-                                                           return event.nodesDelta().masterNodeChanged();
-                                                       }
-                                                   }
-                        );
-                    } else {
-                        listener.onFailure(exp);
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        logger.warn("timed out while retrying. Reason for retry: ", failure);
```

Comment:
Somewhat simpler: ("timed out while retrying [{}] after failure (timeout [{}])", action, failure) . I'm doubting between DEBUG and WARN for this log...